### DISCUSSION TITLE
Artist location

### DISCUSSION THREAD

**Discussion by simonv3** [funmusicplace/mirlo#271] [Ideas]
Make it possible to store an artist's location, and have the artist edit it. 

In the future this could be used to search for artists in a specific area too.

**Comment by bbstilson** [comment]
This seems like a good first issue. The BE seems straightforward, but is there a sense of where this information should display on the FE?

**Reply by bbstilson** [reply]
Actually, looks like this exists. I see routes and UI for it.

**Reply by simonv3** [reply]
@bbstilson I think right now the location we store is a simple string. There's been requests from users that this be slightly more complete. Eg. right now we can't really aggregate based on a location being similar, so we can't really say "find all musicians in ontario."

Honestly I'm not sure what the best way forward is here.

**Reply by bbstilson** [reply]
@simonv3 ah I see. Yeah there are many ways to approach this, but I think the appropriate solution depends more on product direction than technical limitations.

## Keep it Dynamic

It could be kept it as text and encourage artists to put their real location so that people can find them. Autocomplete can be offered to encourage uniformity. Creating a normalized, searchable index on that column would be straightforward, but it puts the onus on the artists and doesn't guarantee conformity. A possibly important pro of this approach is leaving room for creative/unorthodox locations. Some artists might want to use a nickname or "unofficial" name for their city, e.g., "Chi-town" instead of "Chicago".

## Static List
Another alternative would be to populate Mirlo's database with an existing database of countries/states/cities, e.g., https://github.com/dr5hn/countries-states-cities-database. There are a couple cons with this approach:

1) It bloats the db with data that may never be used. That much data is on the order of tens of megabytes, so I don't think there's a big threat to costs or db performance. 
2) It requires artists to choose a city from this list.

The challenge with this approach is existing locations that cannot be cleanly mapped to the canonical set of locations. We could wipe them in the migration by changing the location to null if it cannot be found in the canonical set, but it's unclear how negatively that would impact artists, and how many.

The pro, however, is that search and aggregations can be trivially provided for users and artists alike. I'm personally a fan of having static lists of things, but product needs trump technical preferences.

## Hybrid

A final approach could be a hybrid one where artists can choose from a static list, and opt out if they like. This has more technical complexity both on the backend and UI/UX, but might provide the best of both worlds for artists and users.

**Reply by simonv3** [reply]
Thanks for these options @bbstilson truly appreciate the write-up.

I'm kind of leaning towards the hybrid model right now, and here are some thoughts:

* We need to capture artists real location (well actually country) for shipping reasons anyway. I guess this poses a question on whether this is the `user`s location or the `artist`s. Also this is information that the user already supplies to Stripe (but where the user is from could be very different from where the artist is from).
* I wonder if the current `location` string could be migrated to `locationNickname` and it acts a little bit like a mask. This seems like an simple first step. I think if this value is filled in, it should be what's displayed. If we add real location mapping that's something we can ask artists to fill in during other steps (eg when they set up different shipping locations). 
* I can picture artists wanting to say their location is mars or something like that.

**Reply by bbstilson** [reply]
@simonv3 

Agreed, the hybrid approach sounds best.

> I guess this poses a question on whether this is the users location or the artists.

I think there are few different use-cases related to this feature:

1) As a listener, I want to find artists by location.
2) As an artist, I want people to be able to find me in my location.
3) As an artist, I want to display a different location.

> I wonder if the current location string could be migrated to `locationNickname`.

I like this idea, though I'd suggest something like `locationDisplay` to highlight that it's display text and not a real location. Artists could be given two location input areas:
- "Display Location"
  - This is displayed on the Artist Page to listeners. It takes precedence over "Real Location"
- "Real Location (City/State/Country)"
  - This helps listeners find you by enabling searching by location. If "Display Location" is not populated, then this is displayed on the Artist Page.

Both fields can be optional and mutually inclusive.

As far as the schema design for the actual location, a join table can be used. 

```sql
create table artists_2_locations (
  -- Unique entry per artist...they can't be in two places at once
  artist_id int primary key references artists(id),
  city_id bigint null references cities(id),
  state_id bigint null references states(id),
  country_id bigint null references countries(id),
  -- If the city is defined, then the state should be as well.
  constraint if_city_then_state check ((city_id is null) or (state_id is not null)),
  -- If the state is defined, then the country should be as well.
  constraint if_state_then_country check ((state_id is null) or (country_id is not null))
);
```

A similar table can be made for users. 

Here's a fiddle link with some examples: https://onecompiler.com/postgresql/42yrpssfq

**Reply by simonv3** [reply]
Yeah that all sounds good to me!

**Comment by simonv3** [comment]
Implemented!
